### ISSUE TITLE
Support OIDC private-key-jwt 

### ISSUE THREAD

**Issue by bdecoste** [solo-io/gloo#7744]
### Version

1.12.x

### Is your feature request related to a problem? Please describe.

OIDC client_id + client_secret is less secure than private key jwt (https://oauth.net/private-key-jwt/) for OIDC client authentication. We have a requirement for the latter for increased security. 

### Describe the solution you'd like

Support private key jwt (https://oauth.net/private-key-jwt/) in the OIDC AuthConfig

### Describe alternatives you've considered

less secure client_id + client_secret

**Comment by sam-heilbron** [comment]
This is very similar to https://github.com/solo-io/gloo/issues/8020. When evaluating this work, we should understand the differences and determine what makes sense to build into Gloo Edge. Perhaps it's both, or perhaps we want to encourage one mechanism over anothter.

**Comment by bdecoste** [comment]
We currently only support the authorization code grant, the user has requested that we also support the private-key-jwt grant. In the former, the client_id and client_secret are used to obtain an authorization code which is then used to obtain the tokens. In the latter, the a JWT is used to authenticate the OIDC client (i.e. Gloo) to the tokens

**Comment by sam-heilbron** [comment]
A few items that we'd like to address as part of this work, though not in the existing PRs:
- https://github.com/solo-io/ext-auth-service/pull/644#discussion_r1337631231
- https://github.com/solo-io/ext-auth-service/pull/644#discussion_r1337628173

**Comment by sheidkamp** [comment]
Follow up on:
> * [OIDC code exchange refactor only ext-auth-service#644 (comment)](https://github.com/solo-io/ext-auth-service/pull/644#discussion_r1337631231)
----
in clientautheticator.Exchange:
```
	if e.discovery == nil {
		return nil, ErrDiscoveryNotSet
	}
	d := e.discovery.(*discovery.DiscoveryImpl)
```
We have a discovery component, which asynchronously updates DiscoveryData and then exposes that data. Right now that discovery is shared by more components, it's unclear to me how access is restricted. Are there any concerns about atomic updates leading to race conditions where discovery updates a value (as it does asynchronously) and that causes issues with this getter?
----

"legacy" answer: this is the 'setDiscovery' function:

```
func (iss *IssuerImpl) SetDiscovery(disc discovery.Discovery) {
	disc.Initialize(iss.oauth2Config())
	iss.discovery.Store(disc)
	iss.ClientAuthenticator.SetDiscovery(disc) // Newly added
}
```

and discovery is Stored (and Loaded) in an atomic manner

we are not setting discovery on the clientAuthenticator atomically however. Proposed solution is to configure the clientAuthenticator with a function to get the Discovery instead of storing it locally. This way we can point at the atomic "Load" and ensure we getting/setting safely. Something like https://github.com/solo-io/ext-auth-service/pull/651

Does this approach make sense?

**Comment by sheidkamp** [comment]
> * [OIDC code exchange refactor only ext-auth-service#644 (comment)](https://github.com/solo-io/ext-auth-service/pull/644#discussion_r1337628173)

code:
in clientautheticator.Exchange:
```
	d := e.discovery
	var client *http.Client
	if d.GetClient() != nil {
		client = d.GetClient()
		if d.GetSessionIdHeaderName() != "" {
			client.Transport = discovery.SessionIdSetterTransport(sessionId, d.GetSessionIdHeaderName(), client.Transport)
		}
	} else {
		client = &http.Client{}
		if d.GetSessionIdHeaderName() != "" {
			client.Transport = discovery.SessionIdSetterTransport(sessionId, d.GetSessionIdHeaderName(), http.DefaultTransport)
		}
	}
```
comment:
The discovery client is configured with some extra metadata: https://github.com/solo-io/ext-auth-service/blob/master/pkg/config/oidc/oidc.go#L227. Does the exchanger need to share this client? If so, I think it would be help to clarify why exactly.


I don't think we need to share more than we already are. We do first attempt to get the client from discovery. The client gets created in `IssuerImpl.Discover` and is passed to `discovery.Discover`, where it gets set in the DiscoveryImpl struct. If this client is not defined, that means either `Discover` was not called or no CA Certs were defined. In the first case, that means we are testing and are managing the coordination of components ourselves, and in the second case the logic that is executed looks like (iss.roots is set based on the CA Certs):

```	var httpClient *http.Client
	if iss.roots != nil {
		httpClient = &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{
					RootCAs: iss.roots,
				},
			},
		}
	}
```

so if the certs are not defined, we would have nothing to do, as we are not configuring anything else there.

Additionally, it looks like the parts of discovery that are needed (the SessionIdHeaderName) are being pulled in from the discovery object as needed. So I think we are using the discovery client when possible, and have the discovery information available when we need it if we are not able to reuse the discovery client.

**Comment by sheidkamp** [comment]
This will be available in 1.16.0-beta1 of GlooEE
